src/Stratis.Bitcoin/Consensus/ChainedHeaderTree.cs

```diff
@@ -0,0 +1,536 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using Microsoft.Extensions.Logging;
+using NBitcoin;
+using Stratis.Bitcoin.Base;
+using Stratis.Bitcoin.Configuration.Settings;
+using Stratis.Bitcoin.P2P.Peer;
+using Stratis.Bitcoin.Utilities;
+using Stratis.Bitcoin.Utilities.Extensions;
+
+namespace Stratis.Bitcoin.Consensus
+{
+    public interface IChainedHeaderValidator
+    {
+        void Validate(ChainedHeader chainedHeader);
+    }
+
+    /// <summary>
+    /// A tree structure of <see cref="ChainedHeader "/> elements.
+    /// </summary>
+    /// <remarks>
+    /// A blockchain can have multiple versions of the chain, with only one being the longest chain (or the chain with most proof).
```

Comment:
>A blockchain can have multiple versions of the chain

Not sure what you've meant by that

> with only one being the longest chain

There can be >1 chains of same length and chainwork, just not for a long time. 

>(or the chain with most proof)

Cumulative chainwork